File: Sources/smatrix/Drop.swift
struct Drop {
	var position: Int
	var length: Int
	var characters: [Character]
	var speed: Int

	init(length: Int, speed: Int = 1) {
		self.position = 0
		self.length = length
		self.speed = speed
		self.characters = (0..<length).map { _ in Config.chars.randomElement()! }
	}

	mutating func update() {
		position += speed
	}

	func isActive(in height: Int) -> Bool {
		position < height + length
	}

	func characterAt(row: Int) -> Character? {
		let relativePos = row - position
		guard relativePos >= -length && relativePos < 0 else { return nil }
		let index = relativePos + length
		return characters[index]
	}

	func intensityAt(row: Int) -> Double {
		let relativePos = row - position
		guard relativePos >= -length && relativePos < 0 else { return 0 }
		let index = relativePos + length
		return Double(index + 1) / Double(length)
	}
}

File: src/pmatrix/drop.py
import random
from dataclasses import dataclass

from .config import CHARS, MAX_DROP_LENGTH, MIN_DROP_LENGTH


@dataclass
class Drop:
	position: int
	length: int
	characters: list[str]
	speed: int

	@classmethod
	def create(cls) -> "Drop":
		length = random.randint(MIN_DROP_LENGTH, MAX_DROP_LENGTH)
		speed = random.randint(1, 2)
		characters = [random.choice(CHARS) for _ in range(length)]
		return cls(position=0, length=length, characters=characters, speed=speed)

	def update(self) -> None:
		self.position += self.speed

	def is_active(self, height: int) -> bool:
		return self.position < height + self.length

	def char_at(self, row: int) -> str | None:
		relative_pos = row - self.position
		if not (-self.length <= relative_pos < 0):
			return None
		index = relative_pos + self.length
		return self.characters[index]

	def intensity_at(self, row: int) -> float:
		relative_pos = row - self.position
		if not (-self.length <= relative_pos < 0):
			return 0.0
		index = relative_pos + self.length
		return (index + 1) / self.length
